File: packages/web/src/engine.test.ts
import { describe, it, expect, vi } from "vitest";
import { WaveUnlock } from "./index.js";
import { friendly } from "./denials.js";
import type { UnlockState, WebTransport } from "./types.js";

const config = { gatewayUrl: "https://x/functions/v1", anonKey: "anon", publishableKey: "wave_pub_x", userNumber: "10001" };

function mockTransport(overrides: Partial<WebTransport> = {}): WebTransport & { written?: Uint8Array } {
  const t: WebTransport & { written?: Uint8Array } = {
    connect: vi.fn(async () => {}),
    write: vi.fn(async (p: Uint8Array) => { t.written = p; }),
    disconnect: vi.fn(),
    ...overrides,
  };
  return t;
}

// Sequence of fetch responses: token, then unlock-stream reads.
function mockFetch(streamStatuses: Array<{ status: string; reason?: string }>) {
  let call = 0;
  return vi.fn(async (url: string) => {
    if (url.endsWith("/partner-auth/token")) {
      return new Response(JSON.stringify({ token: "tok", mode: "test" }), { status: 200 });
    }
    const body = streamStatuses[Math.min(call++, streamStatuses.length - 1)];
    return new Response(JSON.stringify(body), { status: 200 });
  }) as unknown as typeof fetch;
}

async function labels(gen: AsyncGenerator<UnlockState>): Promise<string[]> {
  const out: string[] = [];
  for await (const s of gen) out.push(s.kind);
  return out;
}

describe("WaveUnlock engine", () => {
  it("granted sequence", async () => {
    const transport = mockTransport();
    const wave = new WaveUnlock(config, { transport, fetchImpl: mockFetch([{ status: "granted", reason: "[mock] Access Granted" }]) });
    const seq = await labels(wave.unlock({ cloudTimeoutMs: 500, pollIntervalMs: 10 }));
    expect(seq).toEqual(["scanning", "readerFound", "writing", "awaitingConfirmation", "granted"]);
    expect(Array.from(transport.written!)).toEqual([0x01, ...new TextEncoder().encode("10001")]);
  });

  it("denied sequence maps a friendly reason", async () => {
    const wave = new WaveUnlock(config, { transport: mockTransport(), fetchImpl: mockFetch([{ status: "denied", reason: "Client not found" }]) });
    let reason = "";
    const seq: string[] = [];
    for await (const s of wave.unlock({ cloudTimeoutMs: 500, pollIntervalMs: 10 })) {
      seq.push(s.kind);
      if (s.kind === "denied") reason = s.reason;
    }
    expect(seq).toEqual(["scanning", "readerFound", "writing", "awaitingConfirmation", "denied"]);
    expect(reason).toBe("Member not found");
  });

  it("times out when the cloud stays pending", async () => {
    const wave = new WaveUnlock(config, { transport: mockTransport(), fetchImpl: mockFetch([{ status: "pending" }]) });
    const seq = await labels(wave.unlock({ cloudTimeoutMs: 60, pollIntervalMs: 20 }));
    expect(seq[seq.length - 1]).toBe("timedOut");
  });

  it("fails when the reader won't connect", async () => {
    const transport = mockTransport({ connect: vi.fn(async () => { throw new Error("no device"); }) });
    const wave = new WaveUnlock(config, { transport, fetchImpl: mockFetch([{ status: "granted" }]) });
    const seq = await labels(wave.unlock());
    expect(seq).toEqual(["scanning", "failed"]);
  });
});

describe("friendly", () => {
  it("strips mock tag + matches substrings", () => {
    expect(friendly("[mock] Membership expired")).toBe("Membership expired");
    expect(friendly("Blocked by provider : 604 : Client not found")).toBe("Member not found");
    expect(friendly(null)).toBe("Access denied");
  });
});

File: packages/flutter/test/engine_test.dart
import 'dart:typed_data';
import 'package:flutter_test/flutter_test.dart';
import 'package:wave_unlock/wave_unlock.dart';

const config = WaveConfig(
  gatewayUrl: 'https://x/functions/v1',
  anonKey: 'anon',
  publishableKey: 'wave_pub_x',
  userNumber: '10001',
);

/// Fake gateway returning a scripted outcome.
class FakeGateway implements Gateway {
  final Outcome outcome;
  FakeGateway(this.outcome);
  @override
  Future<String> fetchToken() async => 'tok';
  @override
  Future<Outcome> readOutcome(String token) async => outcome;
  @override
  Future<Outcome> awaitOutcome(String token, Duration timeout, {Duration poll = const Duration(milliseconds: 500)}) async => outcome;
}

Future<List<String>> labels(Stream<UnlockState> s) async =>
    [await for (final st in s) st].map((e) => e.label).toList();

void main() {
  final fast = {
    'scanTimeout': const Duration(seconds: 1),
    'cloudTimeout': const Duration(milliseconds: 300),
  };

  test('granted via direct-BLE verdict', () async {
    final t = MockTransport([const EvtReaderFound(-50), const EvtVerdict(true, 'Granted')]);
    final seq = await labels(runUnlock(t, null, config, scanTimeout: fast['scanTimeout']!, cloudTimeout: fast['cloudTimeout']!));
    expect(seq, ['scanning', 'readerFound', 'writing', 'awaitingConfirmation', 'granted']);
    expect(t.writtenPayload, Uint8List.fromList([0x01, ...'10001'.codeUnits]));
  });

  test('denied maps a friendly reason', () async {
    final t = MockTransport([const EvtReaderFound(-40), const EvtVerdict(false, 'Client not found')]);
    final states = [await for (final s in runUnlock(t, null, config, scanTimeout: fast['scanTimeout']!, cloudTimeout: fast['cloudTimeout']!)) s];
    expect(states.map((e) => e.label).toList(), ['scanning', 'readerFound', 'writing', 'awaitingConfirmation', 'denied']);
    expect((states.last as Denied).reason, 'Member not found');
  });

  test('granted via cloud gateway', () async {
    final t = MockTransport([const EvtReaderFound(-50)]);
    final gw = FakeGateway(const Outcome(OutcomeStatus.granted, '[mock] Access Granted'));
    final seq = await labels(runUnlock(t, gw, config, scanTimeout: fast['scanTimeout']!, cloudTimeout: fast['cloudTimeout']!));
    expect(seq, ['scanning', 'readerFound', 'writing', 'awaitingConfirmation', 'granted']);
  });

  test('times out with no verdict and no gateway', () async {
    final t = MockTransport([const EvtReaderFound(-50)]);
    final seq = await labels(runUnlock(t, null, config, scanTimeout: fast['scanTimeout']!, cloudTimeout: const Duration(milliseconds: 80)));
    expect(seq.last, 'timedOut');
  });

  test('too far does not write', () async {
    final t = MockTransport([const EvtReaderFound(-90)]);
    final seq = await labels(runUnlock(t, null, config, scanTimeout: const Duration(milliseconds: 200), cloudTimeout: fast['cloudTimeout']!));
    expect(seq.take(2).toList(), ['scanning', 'tooFar']);
    expect(t.writtenPayload, isNull);
  });

  test('bluetooth off fails', () async {
    final t = MockTransport([const EvtUnavailable(WaveError.bluetoothOff)]);
    final seq = await labels(runUnlock(t, null, config, scanTimeout: fast['scanTimeout']!, cloudTimeout: fast['cloudTimeout']!));
    expect(seq, ['scanning', 'failed']);
  });

  test('friendly strips mock tag + matches substrings', () {
    expect(friendly('[mock] Membership expired'), 'Membership expired');
    expect(friendly('Blocked by provider : 604 : Client not found'), 'Member not found');
    expect(friendly(null), 'Access denied');
    expect(denialTableLength, 14);
  });
}
